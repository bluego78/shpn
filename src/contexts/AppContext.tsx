/* IMPORT NODE MODULES & COMPONENTS*/
import {createContext} from 'react';

/* IMPORT INTERFACES */
import IUser from '../Interfaces/IUser';
import IAppContext from '../Interfaces/IAppContext';

let selectedNationCookie : string | null = localStorage.getItem("selectednat");

// Create the context
const appContext: IAppContext | any = createContext({ 
    isLoading: false,
    filterIsActive:false,
    usersList: [] as Array<IUser>,
    filteredUsersList: [] as Array<IUser>,
    currentPage: 0,
    totalResults:0,
    selectedNations: selectedNationCookie!==null ? selectedNationCookie.split(",") : [],
    modalIsOpened:false,
});

export default appContext;

