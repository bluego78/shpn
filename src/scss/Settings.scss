.nat-container{
    margin-top:1rem;
    display: flex;
    flex-direction: column;

    .nat-row
    {
        display: flex;
        align-items: center;

        input[type=checkbox]
        {
            margin-right:5px;
        }
    }

    .back-button{
        margin-top:30px;
        width:130px;
    }
    
    
}