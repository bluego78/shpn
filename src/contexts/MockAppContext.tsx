/* IMPORT NODE MODULES & COMPONENTS*/
import {useState} from 'react';
import AppContext from './AppContext';

export const MockedContext = ({children}:any)=> {

    const fakeUsers = [
        {
            id: {value:"XX123"},
            login: { username:"theusername" },
            email:"test@example.com",
            name: {first:"John", last:"Doe"},
            nat:"ch",
            picture: { thumbnail:"https://randomuser.me/api/portraits/thumb/men/56.jpg" },
            phone:"+41 76 12345678",
            cell:"+41 76 36958488",
            location: {
                        city:"Chiasso", 
                        country:"Switzerland", 
                        postcode:"6830", 
                        state:"-",
                        street: {name:"C.so San Gottardo", number:"15",}
                    }
        },
        {
            id: {value:"XX456"},
            login: { username:"theusername2" },
            email:"test2@example.com",
            name: {first:"John", last:"Smith"},
            nat:"gb",
            picture: { thumbnail:"https://randomuser.me/api/portraits/thumb/men/55.jpg" },
            phone:"+41 76 987654321",
            cell:"+41 76 98989898",
            location: {
                        city:"London", 
                        country:"Britain", 
                        postcode:"1234", 
                        state:"-",
                        street: {name:"Park Street", number:"11",}
                    }
        }
    ];
    const [appState, setAppState] = useState({...AppContext, usersList: fakeUsers, filteredUsersList: fakeUsers});
    return <AppContext.Provider value={[appState, setAppState]}>{children}</AppContext.Provider>

}
export default MockedContext;


