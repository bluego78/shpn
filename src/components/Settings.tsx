/* IMPORT CSS */
import '../scss/Settings.scss';

/* IMPORT NODE MODULES & COMPONENTS*/
import { useContext, useEffect } from 'react';
import { Link } from 'react-router-dom';

/* IMPORT INTERFACES */
import IUser from '../Interfaces/IUser';

/* IMPORT CONTEXTS */
import AppContext from '../contexts/AppContext';

const Settings = () => {

    // *appContext* contains all the shared values between the components
    const [appContext, setAppContext] = useContext(AppContext);

    //let selectedNationCookie : string | null = localStorage.getItem("selectednat");
    const nationalities : Array<string> = (process.env.REACT_APP_DEFAULT_NATIONALITIES as string).split(",");
    const natNames: Array<string> = (process.env.REACT_APP_DEFAULT_NATIONALITIES_NAMES as string).split(",");

    const setNationalities = (e:React.ChangeEvent<HTMLInputElement>) => {

        //Get the value
        let ckeched:boolean = e.target.checked;
        let value:string = e.target.value;

        //Copy the natlist from the context (the context is updated to the cookie)
        let natList = [...appContext.selectedNations];

        if(ckeched)
        {
            if(!natList.includes(value))
            {
                //add the nat to the list
                natList.push(value);
            }
        }
        else
        {
            if(natList.includes(value))
            {
                //remove the nat from the list
                natList = natList.filter(nat=>nat!==value);
            }
        }

        //update state context selectedNation and clean the search
        setAppContext({...appContext, 
            selectedNations: natList, 
            filterIsActive:false,
            usersList: [] as Array<IUser>, 
            filteredUsersList: [] as Array<IUser>, 
            currentPage: 0,
            totalResults:0
        });

        //set te cookie to a comma separated value of the nationalities
        localStorage.setItem("selectednat",natList.join());
    }

    useEffect(()=>{

        //Empty and hide the search field when go to the settings page
        var searchField: HTMLInputElement = document.getElementById('SearchField') as HTMLInputElement;
        if(searchField){
            searchField.value="";
            searchField?.classList.add("d-none");
        }

    },[])

    return <div className="nat-container">
                <h3>Available Nationalities</h3>
                <p>Please, select the nationalities you want to filter.</p>
                {nationalities.map((nat,i)=>{
                    //verify if is checked in the context & the cookie
                    let checked:boolean = appContext.selectedNations.includes(nat);
                    return <div key={i} className="nat-row">
                        <input id={nat.toLowerCase()} type="checkbox" onChange={(e)=>setNationalities(e)} value={nat} checked={checked}/>{natNames[i]} ({nat})
                        </div>
                })}
                <Link to="/home" className="btn btn-primary back-button">back</Link>
            </div>
}

export default Settings;