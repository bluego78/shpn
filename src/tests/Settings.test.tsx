
import { MockedContextState } from '../contexts/MockAppContext';
import Settings from '../components/Settings';
import renderer from 'react-test-renderer';
import { BrowserRouter as Router } from "react-router-dom";

let component:any;

beforeEach(()=>{
    component = renderer.create(<MockedContextState><Router><Settings /></Router></MockedContextState>);
});

afterEach(()=>{
    component = null;
});

describe('<Settings />', ()=>{

    it('Renders properly', ()=>{
        expect(component.toJSON()).toMatchSnapshot();
    });

    it('Renders the nat container', ()=>{
        expect(component.root.findByProps({className: "nat-container"})).toBeTruthy();
    });

    it('Renders the title', ()=>{
        expect(component.root.findByType('h3').children).toEqual(['Available Nationalities']);
    });

    it('Renders the description', ()=>{
        expect(component.root.findByType('p').children).toEqual(['Please, select the nationalities you want to filter.']);
    });

    it('Renders all the input checkbox', ()=>{
        expect(component.root.findAllByType('input').length).toEqual(process.env.REACT_APP_DEFAULT_NATIONALITIES?.split(",").length);

    });

});