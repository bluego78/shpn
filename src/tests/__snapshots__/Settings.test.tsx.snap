// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Settings /> Renders properly 1`] = `
<div
  className="nat-container"
>
  <h3>
    Available Nationalities
  </h3>
  <p>
    Please, select the nationalities you want to filter.
  </p>
  <div
    className="nat-row"
  >
    <input
      checked={false}
      id="ch"
      onChange={[Function]}
      type="checkbox"
      value="CH"
    />
    Switzerland
     (
    CH
    )
  </div>
  <div
    className="nat-row"
  >
    <input
      checked={false}
      id="es"
      onChange={[Function]}
      type="checkbox"
      value="ES"
    />
    Spain
     (
    ES
    )
  </div>
  <div
    className="nat-row"
  >
    <input
      checked={false}
      id="fr"
      onChange={[Function]}
      type="checkbox"
      value="FR"
    />
    France
     (
    FR
    )
  </div>
  <div
    className="nat-row"
  >
    <input
      checked={false}
      id="gb"
      onChange={[Function]}
      type="checkbox"
      value="GB"
    />
    Britain
     (
    GB
    )
  </div>
  <a
    className="btn btn-primary back-button"
    href="/home"
    onClick={[Function]}
  >
    back
  </a>
</div>
`;
